File: algorithm/acwing/LanQiaoBeiDayOne2023/src/Week1/day3/AcWing113/Solution.java
package Week1.day3.AcWing113;

/**
 * @Author 袁佳林
 * @Description
 * @Date 2023/2/22 21:09
 */
class Solution extends Relation {
    public int[] specialSort(int N) {
        int[] res = new int[N];
        res[0] = 1;
        for(int i = 2;i <= N;i++) {
            // 二分插入点
            int l = 0, r = i - 2;
            while(l < r) {
                int mid = l + ((r - l) >> 1);
                if(compare(i, res[mid])) {
                    r = mid;
                } else {
                    l = mid + 1;
                }
            }

            int j = i - 1;
            if(compare(i,res[l])) { // 当前有解
                // 插入
                while(j > l) {
                    res[j] = res[j - 1];
                    j--;
                }
            }
            res[j] = i;
        }
        return res;
    }
}

File: algorithm/acwing/LanQiaoBeiDayOne2023/src/Week1/day3/AcWing113/Solution.py
#author: 袁佳林
class Solution(object):
    def specialSort(self, N):
        """
        :type N: int
        :rtype: List[int]
        """
        res = [1]
        for i in range(2, N+1):
            l, r = -1, len(res) - 1
            while l < r:
                mid = (l + r + 1) >> 1
                if compare(res[mid],i):
                    l = mid
                else:
                    r = mid - 1
            res.insert(l+1, i)

        return res

